File: lab/econwar_engine/rng.py
"""mulberry32 — the pinned PRNG, ported to match the TS engine bit-for-bit.

All ops are performed on unsigned 32-bit values (masked), which reproduces
JS's int32/uint32/Math.imul semantics exactly.
"""

MASK32 = 0xFFFFFFFF
GOLDEN_RATIO = 0x9E3779B1


def _imul(a: int, b: int) -> int:
    """Equivalent of JS Math.imul on the low 32 bits."""
    return (a * b) & MASK32


class Rng:
    def __init__(self, seed: int) -> None:
        self.a = seed & MASK32

    def next_u32(self) -> int:
        self.a = (self.a + 0x6D2B79F5) & MASK32
        a = self.a
        t = _imul(a ^ (a >> 15), (1 | a) & MASK32)
        t = (((t + _imul(t ^ (t >> 7), (61 | t) & MASK32)) & MASK32) ^ t) & MASK32
        return (t ^ (t >> 14)) & MASK32

    def next_below(self, n: int) -> int:
        if not isinstance(n, int) or n <= 0:
            raise ValueError(f"next_below requires a positive integer, got {n}")
        u = self.next_u32()
        return (u * n) >> 32


def make_rng(seed: int) -> Rng:
    return Rng(seed)


def phase_seed(game_seed: int, phase_index: int) -> int:
    return (game_seed ^ _imul(phase_index + 1, GOLDEN_RATIO)) & MASK32

File: packages/engine/src/rng.ts
/**
 * The one pinned PRNG for the entire project: `mulberry32`.
 *
 * Golden rule #3: deterministic RNG, identical in TS and Python, locked by a
 * committed golden vector (golden/rng_vector.json). No Math.random(), no
 * Date.now() — seed in, identical uint32 stream out.
 *
 * We expose the raw uint32 output (an exact integer, never a fractional float)
 * and integer selection helpers, so the engine honors the "no floats" rule.
 */

const FNV_LIKE_SALT = 0x9e3779b1; // golden-ratio constant for phase-seed mixing

export class Rng {
  private a: number;

  constructor(seed: number) {
    this.a = seed | 0;
  }

  /** Next 32-bit unsigned integer in [0, 2^32). Canonical mulberry32. */
  nextU32(): number {
    this.a |= 0;
    this.a = (this.a + 0x6d2b79f5) | 0;
    let t = Math.imul(this.a ^ (this.a >>> 15), 1 | this.a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return (t ^ (t >>> 14)) >>> 0;
  }

  /**
   * Unbiased integer in [0, n) via 32-bit multiply-shift, computed in BigInt
   * so it stays exact and float-free. n must be a positive integer.
   */
  nextBelow(n: number): number {
    if (!Number.isInteger(n) || n <= 0) {
      throw new Error(`nextBelow requires a positive integer, got ${n}`);
    }
    const u = BigInt(this.nextU32());
    return Number((u * BigInt(n)) >> 32n);
  }
}

export function makeRng(seed: number): Rng {
  return new Rng(seed);
}

/**
 * Derive a stable per-phase seed from the game seed and phase index, so each
 * phase draws an independent-but-reproducible event stream.
 */
export function phaseSeed(gameSeed: number, phaseIndex: number): number {
  return (gameSeed ^ Math.imul(phaseIndex + 1, FNV_LIKE_SALT)) >>> 0;
}
